src/Validator/Environment.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\Session\Validator;
+
+use function is_string;
+use function session_id;
+
+/** @psalm-import-type OptionsArgument from RemoteAddr */
+final class Environment
+{
+    public function __construct(
+        public readonly ?string $userAgent = null,
+        public readonly ?string $remoteAddr = null,
+        public readonly ?string $sessionId = null
+    ) {
+    }
+
+    /** @param OptionsArgument $options */
+    public static function fromGlobals(array $server, array $options = []): self
+    {
+        $userAgent = isset($server['HTTP_USER_AGENT']) && is_string($server['HTTP_USER_AGENT'])
+            ? $server['HTTP_USER_AGENT']
+            : null;
+
+        $remoteAddr = isset($server['REMOTE_ADDR']) && is_string($server['REMOTE_ADDR'])
+            ? $server['REMOTE_ADDR']
+            : null;
+
+        if ($remoteAddr === null || (isset($options['use_proxy']) && $options['use_proxy'])) {
+            $remoteAddr = RemoteAddr::getIpAddress($options, $remoteAddr);
+        }
```

Comment:
Accessing static methods on a validator here is the wrong thing to do - you are making this value object dependent on the validator - `RemoteAddr` needs the `X-ForwardedFor`, maybe `X-ForwardedProto` or whatever - these should be made public properties and simply copied from the given array.

```suggestion
        $forwardedFor = isset($server['X_FORWARDED_FOR']) && is_string($server['X_FORWARDED_FOR'])
            ? $server['X_FORWARDED_FOR']
            : null;
```

The `$options` argument should also be dropped